### ISSUE TITLE
Support for Promises?

### ISSUE THREAD

**Issue by ToucheSir** [natevw/fermata#49]
Having (optional?) support for promises will allow for even more elegant usage with libaries like co. Since fermata is already using es6 features, it wouldn't be too farfetched to either look for a global promise object or expect clients to provide custom promise implementations before using the promise interface.

**Comment by legomind** [comment]
Any update on this?

**Comment by natevw** [comment]
@legomind Not really, I'm hoping to polish up and release the `upcoming` branch soon but I'm not sure if this will make the cut. There's still a lot of unknowns from my perspective:

- what the API syntax for "triggering" a promise to get returned looks like. Do you pass a magical constant in place of the current callback? [this is still the likeliest]
- how to give access to the request-ish object that is *currently* returned from an HTTP method call and would now have to give way to a promise instead
- what the (single) fulfillment value looks like — besides the error parameter, Fermata now returns both a data-or-readable *and* a metadata parameter to callbacks.
- how exactly the approach to using native/library Promise looks like (similarities/differences to similar situation with the Proxy object)

I do think promises are kind of nifty, but to be honest I'm not sold on them either. They appear only a few places in new browser API, a few libraries like jQuery sort of use them a little, while node.js itself doesn't at all… — so it seems like using them instead of the "normal" mechanisms [events/callbacks/etc.] would always be an uphill battle and honestly doesn't seem worth it to me personally.

But clearly there are many people who *are* using promises and would like a better solution for this library. I'll try take another look soon, and would appreciate others to chime in on the questions above, and maybe this can happen despite my lack of a personal "itch" for it.

**Comment by natevw** [comment]
Wanted to write up some recent (design) progress I've made on this recently, especially a potential breakthrough I had the other day on this.

First off, I think at least in first implementation the whole "how do you handle promise library" thing might be solvable by simply having a `fermata.Promise = null` property on the global/module. If you want to use promises, assign your favorite promise class to that property. I think that resolves that worry I've had.

The bigger idea follows below.

Background: I've not been excited about Promise support because one key "trick" Fermata uses is the presence  of a callback function as signal of intent! That is to say `url.path({thing:1})` means "add /path?thing=1" to the URL whereas `url.path({thing:1}, function (e,d,m) { /* … */})` means "send a PATH request with that data in the body". Now PATH is not an HTTP method I've heard of, but the underlying principle is to support any possible path component ("folder") name within a URL while also supporting any possible HTTP method name. So the simple rule is if you pass a callback, it fires off a request, if you don't it simply extends the URL. So for promises I've been thinking you'd need to still pass in some stub/signal function like `url.get(fermata.FAKE_PROMISE_TRIGGERING_CALLBACK)` which just seemed ugly and not worth it.

However my realization is that the Promise API still requires the user to (eventually) provide a callback! So what if, after a user set `fermata.Promise = MyPromiseClass`, we start returning an overridden instance of a Promise object from *every* URL modification/extension? (So the fermata URL "type" essentially becomes a subclass of Promise.) Or at least from the empty `()` call that currently returns a string representation of the URL. Anyway, that Promise subclass/override would intercept the `.then` method (and `catch` assuming that's necessary) with some similar sort of logic as we have in the previous paragraph. If you pass a callback to `.then` it starts an actual HTTP request in motion, otherwise it just extends the path with `/then`.

The example would be something like this:

```
var fermata = require('fermata');
fermata.Promise = global.Promise || require('some-promise-lib');

var baseUrl = fermata.json("http://example.com/api/v1"),
    wordObject = baseUrl.word_list.english.verbs.get;

wordObject.etymology.get().then(function (args) { var data = args[0], metadata = args[1]; /* … */ }, function (err) { /* … * / });
```

Or something like that. (To reiterate, I haven't voluntarily used Promises myself much so maybe there's a better example or I'm missing something that will prevent this from working.)

The basic gist would be combining all three of whatever "normally" gets returned by the access mechanism, the Promise instance needed, and the usual request thing you can use to cancel it. Seems messy but might just work?

**Comment by natevw** [comment]
Oh, I didn't call this out, but worth expounding on here:

Another thing that made me squeamish about promises is that fermata now returns both data and metadata as *two* parameters (in addition to the error parameter which of course has its own mechanism with Promises). But promises can only be resolved with a single value, not two+.

But I think the answer there is to just assume if Promises and the rest of ES6 take off, then you have destructuring anyway to clean it up a bit `var [data, metadata] = args;`. Not ideal but not horrible either?

**Comment by ricxsar** [comment]
Any update of this?

**Comment by natevw** [comment]
@ricxsar No, I haven't gotten much feedback lately and don't use Promises enough myself to judge how useful this would be.

**Comment by natevw** [comment]
Update: we need Promise support for `async`/`await`. See also #65.

**Comment by natevw** [comment]
`await` does work with other objects, since ECMAScript 2017 section 25.5.5.3 [defines AsyncFunctionAwait](https://www.ecma-international.org/ecma-262/8.0/index.html#sec-async-functions-abstract-operations-async-function-await) ± in terms of `Promise.resolve` on await's value, which in turn works on any then-able object:


```
function FakePromise() {}
FakePromise.prototype.then = function (success) {
  setTimeout(function () { success("it worked"); }, 1e3);
};

async function main() {
  console.log(await new FakePromise())
}
main().catch(console.error)
```

So it might be possible to have every URL instance include a `.then` method which I think could use the same "did we get passed a callback?" [instead of string/array/object] as the existing HTTP method `.get`/`.put`/`.del`/etc. does.

```
let baseUrl = fermata.json("http://example.com/api/v1"),
    wordPage = baseUrl.word_list.english.verbs.get
assert(wordPage() === "http://example.com/api/v1/word_list/english/verbs/get")

try {
   let data = await wordPage.etymology.post({sampleUsage:"Get it?"})
   let meta = data[fermata.METADATA_SYMBOL]    // not sure where else to put it…
} catch (err) {
   console.error(err)
}
```
